File: TestLens/testlens-fixture/src/services/UserService.ts
import { MAX_NAME_LENGTH, User, UserId } from "../models/User";
import { UserRepository } from "../repositories/UserRepository";

export interface CreateUserInput {
  id: UserId;
  email: string;
  name: string;
  passwordHash: string;
}

export class UserService {
  constructor(private readonly repo: UserRepository) {}

  createUser(input: CreateUserInput): User {
    if (!input.email.includes("@")) {
      throw new Error("Email must be valid");
    }
    if (input.name.length > MAX_NAME_LENGTH) {
      throw new Error("Name is too long");
    }
    if (this.repo.findByEmail(input.email)) {
      throw new Error("Email already exists");
    }

    const user: User = {
      id: input.id,
      email: input.email,
      name: input.name,
      passwordHash: input.passwordHash
    };
    this.repo.save(user);
    return user;
  }

  getUser(id: UserId): User | null {
    return this.repo.findById(id);
  }

  deleteUser(id: UserId): boolean {
    return this.repo.delete(id);
  }
}

File: bitloops_cli/tests/fixtures/testlens-fixture-rust/src/services/user_service.rs
use crate::models::user::User;
use crate::repositories::user_repository::UserRepository;
use crate::services::auth_service::AuthService;

#[derive(Debug, Default, Clone, Copy)]
pub struct UserService;

impl UserService {
    pub fn create_user(repo: &mut UserRepository, id: u32, email: &str, raw_password: &str) -> User {
        let password_hash = AuthService::hash_password(raw_password);
        let user = User::new(id, email.to_string(), password_hash);
        repo.save(user.clone());
        user
    }

    pub fn authenticate(repo: &UserRepository, email: &str, raw_password: &str) -> bool {
        if let Some(user) = repo.find_by_email(email) {
            return AuthService::verify_password(raw_password, &user.password_hash);
        }
        false
    }
}
